test/e2e/aks_webhooks.go

```diff
@@ -0,0 +1,391 @@
+		Expect(err).NotTo(HaveOccurred())
+		if to.String(ammp.Spec.OsDiskType) == "Ephemeral" {
+			ammp.Spec.OsDiskType = to.StringPtr("Managed")
+		} else {
+			ammp.Spec.OsDiskType = to.StringPtr("Ephemeral")
+		}
+		err = mgmtClient.Update(ctx, ammp)
+		Expect(err).To(HaveOccurred())
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+		By("Validating that we can't update AzureManagedMachinePool.Spec.ScaleSetPriority")
+		ammp, err = getAzureManagedMachinePool(ctx, mgmtClient, mp)
+		Expect(err).NotTo(HaveOccurred())
+		if to.String(ammp.Spec.ScaleSetPriority) == "Regular" {
+			ammp.Spec.ScaleSetPriority = to.StringPtr("Spot")
+		} else {
+			ammp.Spec.ScaleSetPriority = to.StringPtr("Regular")
+		}
+		err = mgmtClient.Update(ctx, ammp)
+		Expect(err).To(HaveOccurred())
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+		By("Validating that we can't update AzureManagedMachinePool.Spec.EnableUltraSSD")
+		ammp, err = getAzureManagedMachinePool(ctx, mgmtClient, mp)
+		Expect(err).NotTo(HaveOccurred())
+		if to.Bool(ammp.Spec.EnableUltraSSD) {
+			ammp.Spec.EnableUltraSSD = to.BoolPtr(false)
+		} else {
+			ammp.Spec.EnableUltraSSD = to.BoolPtr(true)
+		}
+		err = mgmtClient.Update(ctx, ammp)
+		Expect(err).To(HaveOccurred())
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+		By("Validating that we can't update AzureManagedMachinePool.Spec.EnableNodePublicIP")
+		ammp, err = getAzureManagedMachinePool(ctx, mgmtClient, mp)
+		Expect(err).NotTo(HaveOccurred())
+		if to.Bool(ammp.Spec.EnableNodePublicIP) {
+			ammp.Spec.EnableNodePublicIP = to.BoolPtr(false)
+		} else {
+			ammp.Spec.EnableNodePublicIP = to.BoolPtr(true)
+		}
+		err = mgmtClient.Update(ctx, ammp)
+		Expect(err).To(HaveOccurred())
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+		By("Validating that we can't update AzureManagedMachinePool.Spec.NodePublicIPPrefixID")
+		ammp, err = getAzureManagedMachinePool(ctx, mgmtClient, mp)
+		Expect(err).NotTo(HaveOccurred())
+		if to.String(ammp.Spec.NodePublicIPPrefixID) != "" {
+			ammp.Spec.NodePublicIPPrefixID = to.StringPtr(to.String(ammp.Spec.NodePublicIPPrefixID) + "foo")
+		} else {
+			ammp.Spec.NodePublicIPPrefixID = to.StringPtr("foo")
+		}
+		err = mgmtClient.Update(ctx, ammp)
+		Expect(err).To(HaveOccurred())
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+		By("Validating that we can't update AzureManagedMachinePool.Spec.KubeletConfig")
+		ammp, err = getAzureManagedMachinePool(ctx, mgmtClient, mp)
+		Expect(err).NotTo(HaveOccurred())
+		if ammp.Spec.KubeletConfig != nil {
+			ammp.Spec.KubeletConfig.FailSwapOn = to.BoolPtr(!to.Bool(ammp.Spec.KubeletConfig.FailSwapOn))
+		} else {
+			ammp.Spec.KubeletConfig = &infrav1exp.KubeletConfig{
+				FailSwapOn: to.BoolPtr(true),
+			}
+		}
+		err = mgmtClient.Update(ctx, ammp)
+		Expect(err).To(HaveOccurred())
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+		By(fmt.Sprintf("Changing AzureManagedMachinePool %s mode to User", ammp.Name))
```

Comment:
I can see where if the cluster template ever includes more than one System node pool, then this test would have a side effect of changing all but one of them to User node pools. I think the likelihood of the template changing that way and that causing any actual issues is low enough that I doubt it's worth addressing that now though.

Reply:
I think we probably actually do want to anticipate a test scenario where we have (or update to) > 1 system pool. Because changing back and forth is in fact allowed, we should be cool to do that a bunch of times, and this particular test which does that in an excessive/not allowed way supports that.

(So long as we don't do something weird like leave a million kube-system pods around after we do a multiple system pool scenario and then leave around a bunch of orphaned, non-schedulable pods when we return to only 1 system pool — although even that particular weird edge case should still allow tests tests to pass AFAIK)